File: cpp/src/solutions/longest_substring_without_repeating_characters.cpp
// LeetCode 3. Longest Substring Without Repeating Characters

#include "cpp/src/common/common.h"

class Solution {
 public:
  int lengthOfLongestSubstring(string s) {
    vector<int> letter(128, 0);
    int ans = 0;
    for (int l = 0, r = 0; r < s.size(); r++) {
      l = max(l, letter[s[r]]);
      ans = max(ans, r - l + 1);
      letter[s[r]] = r + 1;
    }
    return ans;
  }
};

TEST(lengthOfLongestSubstring, example1) {
  string s = "abcabcbb";
  int ans = 3;
  EXPECT_EQ(Solution().lengthOfLongestSubstring(s), ans);
}

TEST(lengthOfLongestSubstring, example2) {
  string s = "bbbbb";
  int ans = 1;
  EXPECT_EQ(Solution().lengthOfLongestSubstring(s), ans);
}

TEST(lengthOfLongestSubstring, example3) {
  string s = "pwwkew";
  int ans = 3;
  EXPECT_EQ(Solution().lengthOfLongestSubstring(s), ans);
}

TEST(lengthOfLongestSubstring, example4) {
  string s = "";
  int ans = 0;
  EXPECT_EQ(Solution().lengthOfLongestSubstring(s), ans);
}

File: rust/src/solutions/longest_substring_without_repeating_characters.rs
// LeetCode 3. Longest Substring Without Repeating Characters

use super::Solution;

impl Solution {
    pub fn length_of_longest_substring(s: String) -> i32 {
        let mut letter = [0; 128];
        let mut l = 0;
        let mut ans = 0;
        for (r, ch) in s.chars().enumerate() {
            l = std::cmp::max(l, letter[ch as usize]);
            ans = std::cmp::max(ans, r - l + 1);
            letter[ch as usize] = r + 1;
        }
        ans as i32
    }
}

#[cfg(test)]
mod tests {
    use super::Solution;

    #[test]
    fn example1() {
        let s: String = "abcabcbb".into();
        let ans: i32 = 3;
        assert_eq!(Solution::length_of_longest_substring(s), ans);
    }

    #[test]
    fn example2() {
        let s: String = "bbbbb".into();
        let ans: i32 = 1;
        assert_eq!(Solution::length_of_longest_substring(s), ans);
    }

    #[test]
    fn example3() {
        let s: String = "pwwkew".into();
        let ans: i32 = 3;
        assert_eq!(Solution::length_of_longest_substring(s), ans);
    }

    #[test]
    fn example4() {
        let s: String = "".into();
        let ans: i32 = 0;
        assert_eq!(Solution::length_of_longest_substring(s), ans);
    }
}
